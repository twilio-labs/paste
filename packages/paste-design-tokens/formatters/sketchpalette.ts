import * as color from 'color';
import * as Immutable from 'immutable';
import {ImmutableStyleMap} from 'theo';

interface SketchPaletteColor {
  name: string;
  red: number;
  green: number;
  blue: number;
  alpha: number;
}

interface ColorRGB {
  model: string;
  color: string[];
  valpha: string;
}

const convertToSketchPaletteColor = (input: string): {[key: string]: number} => {
  // @ts-ignore FIXME
  const convertedColor: ColorRGB = color.rgb(input);

  return {
    red: parseInt(convertedColor.color[0], 10) / 255,
    green: parseInt(convertedColor.color[1], 10) / 255,
    blue: parseInt(convertedColor.color[2], 10) / 255,
    alpha: parseInt(convertedColor.valpha, 10),
  };
};

const convertColorArray = (result: ImmutableStyleMap): SketchPaletteColor[] =>
  result
    .get('props')
    .sortBy(prop => {
      if (prop !== undefined) {
        return prop.get('name');
      }
    })
    .map(prop => {
      if (prop === undefined) return Immutable.Map();
      const value = prop.get('value');
      const name = prop.get('name');
      if (typeof value !== 'string') {
        throw new TypeError(`[SketchPallete.ts] Value must be of type "string".`);
      }
      return Immutable.Map({name}).merge(convertToSketchPaletteColor(value));
    })
    .toJS();

export const sketchpaletteTokenFormat = (result: ImmutableStyleMap): string => {
  return JSON.stringify({
    compatibleVersion: '2.0',
    pluginVersion: '2.21',
    colors: convertColorArray(result),
    gradients: [],
    images: [],
  });
};
