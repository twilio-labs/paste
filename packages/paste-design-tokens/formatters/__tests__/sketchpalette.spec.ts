import * as theo from 'theo';
import {resolve} from 'path';
import {sketchpaletteTokenFormat} from '../sketchpalette';

theo.registerFormat('sketchpalette', sketchpaletteTokenFormat);

describe('sketchPaletteTokenFormatter', () => {
  it('should return sketch palette color formatted tokens', () => {
    return theo
      .convert({
        transform: {
          type: 'web',
          file: resolve(__dirname, '../__fixtures__/colors.yml'),
        },
        format: {
          // @ts-ignore Theo isn't typed for custom format types
          type: 'sketchpalette',
        },
      })
      .then((sketchPalette: string) => {
        return expect(sketchPalette).toMatchSnapshot();
      })
      .catch((error: string) => {
        console.log(`Something  went wrong: ${error}`);
        throw new Error('[sketchpalette test]: should return sketch palette color formatted tokens threw');
      });
  });
});
