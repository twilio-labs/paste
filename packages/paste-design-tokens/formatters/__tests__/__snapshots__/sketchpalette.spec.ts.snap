// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sketchPaletteTokenFormatter should return sketch palette color formatted tokens 1`] = `"{"compatibleVersion":"2.0","pluginVersion":"2.21","colors":[{"name":"color_background_body","red":1,"green":1,"blue":1,"alpha":1},{"name":"color_background_brand_highlight","red":0.9490196078431372,"green":0.1843137254901961,"blue":0.27450980392156865,"alpha":1},{"name":"color_background_brand_highlight_dark","red":0.9215686274509803,"green":0.058823529411764705,"blue":0.1607843137254902,"alpha":1},{"name":"color_background_brand_highlight_darker","red":0.7607843137254902,"green":0.1450980392156863,"blue":0.21568627450980393,"alpha":1},{"name":"color_background_brand_primary","red":0.050980392156862744,"green":0.07058823529411765,"blue":0.16862745098039217,"alpha":1},{"name":"color_background_default","red":0.9647058823529412,"green":0.9647058823529412,"blue":0.9647058823529412,"alpha":1},{"name":"color_background_element_selection","red":0.9254901960784314,"green":0.7333333333333333,"blue":0.7568627450980392,"alpha":1},{"name":"color_background_error","red":0.807843137254902,"green":0.3137254901960784,"blue":0.37254901960784315,"alpha":1},{"name":"color_background_error_dark","red":0.6823529411764706,"green":0.12941176470588237,"blue":0.19215686274509805,"alpha":1},{"name":"color_background_error_light","red":0.7607843137254902,"green":0.1450980392156863,"blue":0.21568627450980393,"alpha":1},{"name":"color_background_row_active","red":0.9019607843137255,"green":0.9019607843137255,"blue":0.9019607843137255,"alpha":1},{"name":"color_background_row_focus","red":0.8392156862745098,"green":0.8392156862745098,"blue":0.8392156862745098,"alpha":1},{"name":"color_background_row_hover","red":0.8392156862745098,"green":0.8392156862745098,"blue":0.8392156862745098,"alpha":1},{"name":"color_background_row_selected","red":0.9019607843137255,"green":0.9019607843137255,"blue":0.9019607843137255,"alpha":1},{"name":"color_background_row_striped","red":0.9215686274509803,"green":0.9215686274509803,"blue":0.9215686274509803,"alpha":1},{"name":"color_background_success","red":0.39215686274509803,"green":0.8196078431372549,"blue":0.5411764705882353,"alpha":1},{"name":"color_background_success_dark","red":0.3137254901960784,"green":0.6549019607843137,"blue":0.43137254901960786,"alpha":1},{"name":"color_background_success_light","red":0.6352941176470588,"green":0.8901960784313725,"blue":0.7215686274509804,"alpha":1},{"name":"color_background_warning","red":0.984313725490196,"green":0.8156862745098039,"blue":0.5647058823529412,"alpha":1},{"name":"color_background_warning_dark","red":0.7803921568627451,"green":0.5529411764705883,"blue":0.2196078431372549,"alpha":1},{"name":"color_background_warning_light","red":0.9764705882352941,"green":0.6941176470588235,"blue":0.2784313725490196,"alpha":1},{"name":"color_brand_highlight","red":0.9490196078431372,"green":0.1843137254901961,"blue":0.27450980392156865,"alpha":1},{"name":"color_brand_highlight_dark","red":0.9215686274509803,"green":0.058823529411764705,"blue":0.1607843137254902,"alpha":1},{"name":"color_brand_primary","red":0.050980392156862744,"green":0.07058823529411765,"blue":0.16862745098039217,"alpha":1},{"name":"color_gray_1","red":1,"green":1,"blue":1,"alpha":1},{"name":"color_gray_10","red":0.13333333333333333,"green":0.13333333333333333,"blue":0.13333333333333333,"alpha":1},{"name":"color_gray_2","red":0.9647058823529412,"green":0.9647058823529412,"blue":0.9647058823529412,"alpha":1},{"name":"color_gray_3","red":0.9215686274509803,"green":0.9215686274509803,"blue":0.9215686274509803,"alpha":1},{"name":"color_gray_4","red":0.9019607843137255,"green":0.9019607843137255,"blue":0.9019607843137255,"alpha":1},{"name":"color_gray_5","red":0.8392156862745098,"green":0.8392156862745098,"blue":0.8392156862745098,"alpha":1},{"name":"color_gray_6","red":0.8,"green":0.8,"blue":0.8,"alpha":1},{"name":"color_gray_7","red":0.6,"green":0.6,"blue":0.6,"alpha":1},{"name":"color_gray_8","red":0.4,"green":0.4,"blue":0.4,"alpha":1},{"name":"color_gray_9","red":0.26666666666666666,"green":0.26666666666666666,"blue":0.26666666666666666,"alpha":1},{"name":"color_text_brand_highlight","red":0.9490196078431372,"green":0.1843137254901961,"blue":0.27450980392156865,"alpha":1},{"name":"color_text_brand_highlight_dark","red":0.9215686274509803,"green":0.058823529411764705,"blue":0.1607843137254902,"alpha":1},{"name":"color_text_brand_primary","red":1,"green":1,"blue":1,"alpha":1},{"name":"color_text_default","red":0.26666666666666666,"green":0.26666666666666666,"blue":0.26666666666666666,"alpha":1},{"name":"color_text_error","red":0.7607843137254902,"green":0.1450980392156863,"blue":0.21568627450980393,"alpha":1},{"name":"color_text_icon_default","red":0.26666666666666666,"green":0.26666666666666666,"blue":0.26666666666666666,"alpha":1},{"name":"color_text_inverse","red":1,"green":1,"blue":1,"alpha":1},{"name":"color_text_inverse_weak","red":0.9215686274509803,"green":0.9215686274509803,"blue":0.9215686274509803,"alpha":1},{"name":"color_text_label","red":0.26666666666666666,"green":0.26666666666666666,"blue":0.26666666666666666,"alpha":1},{"name":"color_text_link","red":0,"green":0.4392156862745098,"blue":0.8,"alpha":1},{"name":"color_text_link_active","red":0,"green":0.34901960784313724,"blue":0.6392156862745098,"alpha":1},{"name":"color_text_link_destructive","red":0.7607843137254902,"green":0.1450980392156863,"blue":0.21568627450980393,"alpha":1},{"name":"color_text_link_destructive_active","red":0.6078431372549019,"green":0.11372549019607843,"blue":0.17254901960784313,"alpha":1},{"name":"color_text_link_destructive_disabled","red":0.9254901960784314,"green":0.7333333333333333,"blue":0.7568627450980392,"alpha":1},{"name":"color_text_link_destructive_focus","red":0.6823529411764706,"green":0.12941176470588237,"blue":0.19215686274509805,"alpha":1},{"name":"color_text_link_destructive_hover","red":0.6823529411764706,"green":0.12941176470588237,"blue":0.19215686274509805,"alpha":1},{"name":"color_text_link_disabled","red":0.4,"green":0.6627450980392157,"blue":0.8784313725490196,"alpha":1},{"name":"color_text_link_focus","red":0,"green":0.3803921568627451,"blue":0.7529411764705882,"alpha":1},{"name":"color_text_link_hover","red":0,"green":0.3803921568627451,"blue":0.7529411764705882,"alpha":1},{"name":"color_text_placeholder","red":0.4,"green":0.4,"blue":0.4,"alpha":1},{"name":"color_text_success","red":0.39215686274509803,"green":0.8196078431372549,"blue":0.5411764705882353,"alpha":1},{"name":"color_text_warning","red":0.9764705882352941,"green":0.6941176470588235,"blue":0.2784313725490196,"alpha":1},{"name":"color_text_weak","red":0.9019607843137255,"green":0.9019607843137255,"blue":0.9019607843137255,"alpha":1}],"gradients":[],"images":[]}"`;
